### ISSUE TITLE
Support startup dependencies between containers on the same Pod

### ISSUE THREAD

**Issue by nmittler** [kubernetes/kubernetes#65502]
/kind feature
/sig node 

This is a proposed kubernetes feature to address an issue we're seeing on Istio. I'm definitely open for alternative approaches, but I feel that this may be a generally useful feature for Kubernetes.

Background: 

Istio pods have a sidecar container (called "istio-proxy") which runs an Envoy proxy. All traffic in and out of the pod is redirected (via iptables) through Envoy. Kubernetes starts both the application and istio-proxy at the same time, and as soon as the application shows "healthy", it can start receiving traffic. However, if Envoy has not yet received its initial configuration from Pilot (the configuration service), the service will experience temporary traffic loss until Envoy is configured. 

We're currently recommending that developers solve this problem themselves by running a startup script on their application container which delays application startup until Envoy has received its initial configuration.  However, this is a bit of a hack and requires changes to every one of the developer's containers.

Proposal:

In my mind, this situation would be better if we could make the startup of one container on the Pod dependent on a healthy status of another container.  This way, we can provide an Envoy health check that confirms that it's received configuration from Pilot, thus triggering the startup of the application container automatically.  No custom scripts, or manual steps at all.

**Comment by nmittler** [comment]
@louiscryan @PiotrSikora @costinm @rshriram

**Comment by smarterclayton** [comment]
@kubernetes/sig-apps-feature-requests

**Comment by Lion-Wei** [comment]
I love this idea, seems like a big feature. : )

**Comment by ant31** [comment]
To be sure I understand: 
the status of the pod is `1/2`, the first container is healthy and start receiving traffic, while the second one (istio-proxy) is not yet marked as ready/ healthy?

**Comment by nmittler** [comment]
@ant31 Yes, that's correct.

**Comment by ant31** [comment]
Instead of startup dependencies between container inside a pod,  what about a probe for the whole pod? 

it can happen to have mutual-dependencies, in your usecase I think both should be marked as unhealthy until they are both healthy. 
Not sure if it simplifies the implementation but maybe a single field can capture this behavior (groupProbes):

```yaml
spec:
  groupProbes: true   
  containers:
  - name: backend-api
    image: quay.io/myns/backend
    args:
    - /server
    livenessProbe:
      httpGet:
        path: /healthz
        port: 8080

 - name: istio-proxy
    image: istio/istio-proxy
    livenessProbe:
       ....

```
If the groupProbes is set to true, the condition for a container to be marked healthy is:
 - `My probes are Green AND other containers' Probes are Green`
This way the pod would stay `0/2` until both are ready to serve and then directly switch to `2/2` 

(note: I'd like to avoid as much as possible dependencies and startup order)

**Comment by nmittler** [comment]
@ant31 yes, I think aggregated health probes would work as well. So long as traffic isn't being routed to the pod until all containers are "healthy", startup timing shouldn't matter for us.

**Comment by PiotrSikora** [comment]
@ant31 I think that the aggregated health check is also a good idea and might be even a preferred way of checking the health status of a workload with a sidecar, however, I don't think it solves the dependency issue, since all outgoing connections initiated by the workload will be failing before the proxy container becomes healthy, which might result in workload prematurely exiting/crashing/being misconfigured.

**Comment by ant31** [comment]
>   which might result in workload prematurely exiting/crashing/being misconfigured.

Yes, it will crash-loop until the envoy proxy is up, which is perfectly fine: it's not going to be stalled. 
Things with hard dependencies exist, but they should be solved outside kubernetes core in my opinion (like with startup-scripts or init-containers)

I propose to move forward on the aggregated-health-probes proposal. It will solve the described use case for istio, and I think it's globally useful for any kind of workload using a sidecar. 
@nmittler are you ok?

**Comment by nmittler** [comment]
> are you ok?
@ant31 Agreed that aggregated health probes are a generally useful feature and worth implementing. 

Applications with a hard dependency between containers will likely encounter a startup delay for the pod, as @PiotrSikora suggested. I suspect the added delay of crash looping will be acceptable for most applications (especially if the pod won't be receiving traffic).  Let's move forward with aggregated health probes .... if the startup delay is still a problem, we can continue the discussion then. Sound reasonable?

**Comment by omkensey** [comment]
Does the 1.11 beta feature "PodReadinessGates" solve all or most of this?  It looks like work in a similar vein, at least:  https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate

**Comment by nmittler** [comment]
@omkensey that may solve the problem, I'm not quite sure yet.

The definition of the pod phase *Running* is `The Pod has been bound to a node, and all of the Containers have been created. At least one Container is still running, or is in the process of starting or restarting.`

I suspect that once a pod is running, it will start receiving traffic ... even if only one of the containers has been successfully started?  If so, that may not work for this case. We would need the pod to only transition to `Running` after *all* containers are successfully started. Is that achievable with pod-readiness-gates?

**Comment by omkensey** [comment]
It appears to be exactly what PodReadinessGates is for.  Looking at it in more depth last night, it looks like to use PRG it’s expected that something is actually going to be monitoring pods as they come up, and injecting the PRG condition(s) in their status fields, so code would need to be written to do that — but code would need to be written anyway, and using PRG I think the the status injector code could be a controller running in the cluster rather than needing to merged into the mainline Kubernetes code.

**Comment by nmittler** [comment]
> so code would need to be written to do that — but code would need to be written anyway

IIUC, the idea of aggregated health probes would not require any custom code (assuming each container has a means for performing health checks). Or am I missing something?

**Comment by nmittler** [comment]
@omkensey @ant31 After taking another look at [the probe docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes), I see:

`A Pod is considered ready when all of its Containers are ready.`

That may be all we need actually.  If I read that correctly, giving the Envoy container a readiness probe would prevent traffic from being sent to the pod until that container returns `OK`.

**Comment by ant31** [comment]
I still think that the cases you describe should be handle at the application level.

> I've a rather hacky work around, and that is we add a lifecycle hook to istio-proxy, which blocks the shutdown until there are no other tcp listeners (thus, half heartedly ensuring our other service has terminated before envoy starts to terminate).

I don't see this as particularly 'hacky'. There's already mechanics to block the shutdown, the application should know when it can shutdown or not, maybe istio-proxy could embed this check.

**Comment by nmittler** [comment]
@ant31 Possibly, however since the apps are in separate containers, it's not clear to me how istio-proxy would obtain the running state of the application container.  I feel that we're treating Istio's case as a one-off, but I'm not sure it is. It would seem that startup/shutdown dependencies would be a fairly common scenario with multi-container pods. Having k8s handle this would be a powerful feature. I guess I'm not sure I understand the resistance here.

**Comment by uromahn** [comment]
+1 on dependency handling / startup/shutdown sequencing between containers in a pod. We are facing a similar issue here and would love a feature like this.

**Comment by nmittler** [comment]
@ant31 

As suggested, I've started a thread on sigg-apps: https://groups.google.com/forum/#!topic/kubernetes-sig-apps/hEOdsgy4U1Y

FYI @louiscryan @PiotrSikora @costinm @rshriram

**Comment by nmittler** [comment]
@ant31 I haven't had any replies regarding the sig-apps thread.  Did I post to the correct forum? Any way that we can nudge the appropriate folks?

**Comment by seans3** [comment]
Please check out **readiness gates**: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md

**Comment by bjhaid** [comment]
@seans3 readiness gates does not address shutdown behavior...

**Comment by nmittler** [comment]
@seans3 @ant31 can we get some traction on this? From the number of folks chiming in here, it seems pretty clear that there's something here worthy of further discussion.

**Comment by uromahn** [comment]
Just to add some flavor to my previous comment and +1 above:
We are currently running a rather large K8s cluster with way over 1000 pods. Most of our pods have multiple sidecars running in them. While for most sidecars it is irrelevant when they start/stop, there are some where it is very important that they start **before** the main container and for others it is important they stop **after** the main container.
To provide two specific examples:

1. we use a proxy sidecar where all network traffic from and to the main service container routes through. For outbound requests, our proxy does client-side load balancing. We have services that need to make requests to other services immediately at startup to configure itself. If the network proxy sidecar is not up before the service makes that call, it will get a 503 back. So, we had to implement retry logic in the service which "waits" until the proxy sidecar is up. It would be much better if the service would not have to implement such "infrastructure concerns" and instead it could be handled by the infrastructure layer, aka K8s.
This use case is similar to the reported Istio sidecar.

1. we have a log-aggregator sidecar that reads all local log messages of the main container and sends them to a message queue which then ships them to our monitoring/reporting systems. If the log-aggregator sidecar stops **before** the main service container, we potentially lose important log messages and metrics.

**Comment by Lion-Wei** [comment]
I think pod ready++ is for custom defined pod condition, try to make sure pod is working. 
And this proposal aims to specify a startup dependencies between containers, like I have two container in one pod, and I want container2 start first, and after container2 works well, then container1 can start. 
I think pod ready++ can't cover this use case.

@nmittler Am I understand right?

**Comment by nmittler** [comment]
@Lion-Wei Yes, partly. The startup issue that you describe is half of the problem.

**Startup**
Need ability to start one container only after another container has started and become ready. An example of this use case is in Istio ... our sidecar Envoy proxy needs to be up and running before the application container can send traffic. Depending on the delay and how the particular application behaves, the application container may end up crash looping if it can't reach its outbound connections at startup.

**Shutdown**
Need the ability to shutdown one container only after another container terminates. In Istio, we ran into issues where the application needs to communicate to other services that it's going away, but the message is never sent because the Envoy sidecar has already exited.

**Comment by Lion-Wei** [comment]
I think there are things to decide, like:
1. How to make sure the first one have been up and running, maybe using readiness probe.
2. Where to specify the startup sequence, or specify a dependency. 

I think this is a good idea, maybe we can send a initial proposal in mail list, to get more thoughts.

**Comment by nmittler** [comment]
@Lion-Wei 

> maybe we can send a initial proposal in mail list, to get more thoughts.

I actually already did.  See https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-417018265.  I haven't gotten any responses, however.

**Comment by uromahn** [comment]
Just added a reply to the comment which is actually a question. Just to make sure I reach someone who is intimately familiar with the k8s code base, I am looking for some pointer to the code where the actual container creation according to the pod spec is happening in the kubelet.

**Comment by Lion-Wei** [comment]
@uromahn Hi, I think this is what you are looking for. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L578

**Comment by nmittler** [comment]
@rshriram I'm bringing our conversation back here from https://github.com/kubernetes/community/pull/2148#issuecomment-422203844.  I don't want to muddy their discussion :).

**From @rshriram:**
> We want the sidecar to shutdown before the application (not other way around), so that the sidecar envoy no longer receives traffic and forwards to the app

**My reply:**
You raise a good point. But I'm not sure that's quite sufficient either.

For inbound traffic, you're right ... we would want to start rejecting requests bound for the application. However, the application also needs to be able to send outbound requests until it's shutdown (see https://github.com/istio/istio/issues/7136).  So the life-cycles of the application and sidecar containers are a bit more intertwined.

I think the desired behavior would be for the sidecar to receive an event that the pod is shutting down and immediately put Envoy into lameduck mode, where it will reject inbound requests. Then the application container receives SIGTERM (but is still allowed to send outbound requests via Envoy). After the application container is terminated, the sidecar container receives SIGTERM.

I'm not sure all of these pieces exist today, but I think this is probably where we want to be.

**Comment by nmittler** [comment]
@rshriram @PiotrSikora The proposal above would require the ability to put Envoy into a lameduck mode where it immediately starts rejecting inbound traffic. Assuming we have that, I think we can get the behavior we want by making the sidecar container shutdown **after** the application container.  Here's the shutdown sequence:

**Deployment time**

Webhook injection adds a [PreStop hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) to the application container that makes an HTTP request to the Envoy Admin interface telling it to enter lameduck mode.

**Pod Shutdown**

1) PreStop hook executes and Envoy enters lameduck mode
2) k8s shuts down the application container (at this point, no more inbound requests should be received, making application shutdown simpler/faster). If application needs to make outbound requests, it still can because Envoy is still running.
3) k8s shuts down the Envoy sidecar

WDYT?

**Comment by enisoc** [comment]
Building on https://github.com/kubernetes/community/pull/2148#issuecomment-422137240, what if we amend the "shutdown ordering" property to this:

* Shutdown ordering: When a Pod goes down gracefully, we will first execute preStop hooks on all normal containers *and all sidecar containers*. After that, we will signal all normal containers to terminate. Only after all normal containers have terminated will we signal sidecar containers to terminate. The deletion grace period will still apply, after which both normal and sidecar containers are killed.

You could then trigger Envoy to enter lameduck in the sidecar's preStop hook, instead of having to inject it into the normal container.

**Comment by nmittler** [comment]
@enisoc that would be perfect ... I was actually going to suggest that, but I was concerned about getting push-back. Big +1 to this!

**Comment by uromahn** [comment]
@enisoc and @nmittler in general I think the proposed solution is good and would definitely solve the Istio use case, I don't think it could solve the additional issue we are facing of "startup" availability.

In our case, the service is trying to make an outbound call to a config service before its endpoint becomes ready. But if the service makes that call before the istio-proxy is ready, the call will fail with a 503. To overcome this, we now will have to implement retries in our service which we'd rather not have there.

I have also thought about a possible solution which would extend the above proposal but unfortunately, I haven't found the time to write this down.

**Comment by nmittler** [comment]
@uromahn That's correct, I only addressed shutdown in https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-422592567) because that is what is being discussed in https://github.com/kubernetes/community/pull/2148#issuecomment-422203844.

For startup, we've implemented [readiness probes for istio-proxy](https://github.com/istio/istio/pull/7465), which is part of the solution.  To solve the problem we need something like what is proposed [here](https://github.com/kubernetes/community/pull/2148#issuecomment-422137240).

**Comment by nmittler** [comment]
@enisoc So IIUC, we need 3 separate changes to k8s:

**Startup**
- Application containers started only after all sidecars go "ready"

**Shutdown**
- [Sidecars shutdown after application containers](https://github.com/kubernetes/community/pull/2148), which is already in progress
- Revised shutdown ordering, such that all `preStop` hooks are executed before any containers shut down.

Does that sound about right? What are the next steps?

**Comment by barney-s** [comment]
Issues : 
1. Pod receives traffic before envoy container is ready
2. Pod receives traffic when main container is in the process of shutting down
3. main container's outbound traffic fails during prestop hook if envoy container terminates sooner
4. (maybe issue) is there a race condition where the outbound connection can  bypass envoy proxy when both containers are brought up in parallel.

@ant31's proposal for `groupProbes`, would solve 1,2 (shutting down main container will fail readiness check)

for solving 3, requiring envoy to wait till all outbound connections terminate is not sufficient because the main container may open new outbound connections in the last stages. 

for solving 4, we need to ensure istio is healthy before main containers start.

So though sequencing is not an ideal thing and introduces brittleness, it seems more and more inevitable.

**Comment by nmittler** [comment]
@barney-s There's actually another issue:

5. Application container tries to contact another service at startup, but can't because it relies on the Envoy sidecar being up and configured. This can result in the application eventually crash looping, delaying application startup.

We saw this happen on Istio when [attaching a sidecar to an NGINX ingress](https://github.com/istio/istio/issues/3533#issuecomment-366261085). What we really want to happen is that the application container doesn't even get started until the sidecar is ready.

**Comment by barney-s** [comment]
> I'm not sure if there is a major concern here. The pod should already stop receiving traffic, so the only outbound connections would be from processing existing requests. The application container will be getting a SIGTERM shortly anyway, so those outbound requests are doomed to failure.

I was thinking more like a flush() but a service level.
Assume the container is giving up some distributed locks or sending an iam shutting down signal to an external service.

**Comment by nmittler** [comment]
@barney-s 

>> I'm not sure if there is a major concern here. The pod should already stop receiving traffic, so the only outbound connections would be from processing existing requests. The application container will be getting a SIGTERM shortly anyway, so those outbound requests are doomed to failure.

> I was thinking more like a flush() but a service level.
Assume the container is giving up some distributed locks or sending an iam shutting down signal to an external service.

Sure... separating the `preStop` hooks so they run before any containers are terminated should allow the application to perform any flush-like activities.  Is that what you were thinking?

**Comment by ArseniiPetrovich** [comment]
Hi, guys. Any progress on that issue?

**Comment by nmittler** [comment]
We've merged the requirements for this into https://github.com/kubernetes/community/pull/2148#issuecomment-422134989

**Comment by nmittler** [comment]
@enisoc looks like the main thrust of this is being tracked by https://github.com/kubernetes/enhancements/issues/753.  However, I'm not sure it covers the shutdown ordering that we had discussed previously  [here](https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-422599429). Do you know if anyone is looking into that separately?

**Comment by Joseph-Irving** [comment]
@nmittler yeah shutdown ordering is being covered, there was still a bit of debate about the pre-stop hooks but the plan is definitely for sidecars to shutdown after the containers have exited.

**Comment by luksa** [comment]
Note that the sidecar container proposal (https://github.com/kubernetes/enhancements/issues/753) won't solve https://github.com/istio/cni/issues/77. 

IMHO, Envoy should be up and handling the pod's traffic before any of the app's init-containers run. It's reasonable to assume that an application developer might need the services provided by Istio in their init containers also, not just in the pod's main container(s).

**Comment by nmittler** [comment]
@Joseph-Irving 

> @nmittler yeah shutdown ordering is being covered, there was still a bit of debate about the pre-stop hooks but the plan is definitely for sidecars to shutdown after the containers have exited. (currently)

Sounds good ... Just looked at the latest proposal and it seems that this is the piece that I was interested in:

- Pre-stop hooks sent to sidecars before non sidecar containers

This will allow us to lame-duck an Envoy running as a sidecar in the pre-stop hook. That way it will stop accepting new connections while we wait for the application to shutdown gracefully.

**Comment by nmittler** [comment]
@luksa 

> Note that the sidecar container proposal ([kubernetes/enhancements#753](https://github.com/kubernetes/enhancements/issues/753)) won't solve [istio/cni#77](https://github.com/istio/cni/issues/77).

Would running Envoy as a CNI plugin resolve this issue?

**Comment by luksa** [comment]
@nmittler Yes, it would, since the CNI plugin would ensure that Envoy is up before any of the pod's containers are allowed to start.

**Comment by nmittler** [comment]
@luksa I see you're already on it ;) https://github.com/istio/istio/issues/11366

**Comment by kerneltime** [comment]
I have a different use case where a side car container creates a fuse filesystem that is consumed by the application container. The application container should not run the script until the side car is ready. The side car needs to run for as long as the application pod and should be stopped after the application pod. We would like the application container to not host custom logic for detecting the readiness of the mount. We don not route traffic to the pod but process data inside it. The pod level readiness of each container does not help the cross container dependency.

**Comment by kerneltime** [comment]
If Pods are modeled as VM, we are missing what essentially is a custom init/systemd script to control startup and shutdown. Maybe this a missing layer in the Pod and should not be part of the kubernetes code base itself. It could be implemented as a bidirectional FS across containers or a RPC.. 
This does need to be solved once and for all pods and containers within it.

**Comment by bpownow** [comment]
just to add another use case, we're struggling with pods that have several containers (e.g. stats, logging, envoy, and service containers). in the case of crons, whenever envoy and the service start up before the rest of the containers, the pod starts taking traffic before stats and logging can catch up).

**Comment by rdsubhas** [comment]
There is also a simpler solution.

* Similar to how StatefulSets provide `podManagementPolicy` – to have a `containerManagementPolicy` attribute within a pod, with exact same attributes – Ordered or Parallel, with default being Parallel as it is now
* This simply means that, since containers are an array already, they can be started in sequence (and shutdown in reverse order) instead of in parallel
* It's similar to initContainers, except they don't run to completion

It would solve most of our issues. When we say container ordering, most of our pods are running 3-4 containers, and not necessarily 10s or 20s of them. So we can afford to flatten the dependency graph and simply use sequential startup ordering.

**Comment by rdsubhas** [comment]
To be frank, this issue is _really bad_ on production.

* First, `istio-init` will setup iptables rules to route all traffic through istio sidecar
* Then, both application & istio-sidecar start together
* And already application starts sending traffic on boot, even before istio data plane is listening on the port

It might be normal to assume that, well for `Deployment` resources, there will be a few restarts, or occasional `CrashLoopBackoff`s and then it will eventually be healthy. It's bad but tolerable.

But this completely fails for `Job` resources. Most of our Jobs are designed to be one-off, and we _don't_ want them to be re-started upon failure. For Jobs, the application container immediately fails, and that's it – the Job is marked as permanently failed. We see no other way around this for one-off non-idempotent jobs.

The only way we can sidestep this is to either disable sidecar for that specific resource, but that messes up other cluster settings such as mTLS and service-to-service authorization. The other way we're doing right now is to add a random `sleep 15` to the pre-start script, and that's also basically shooting in the dark.

Given kubernetes is almost the de-facto container orchestrator standard, this issue really needs some love...

**Comment by Thoro** [comment]
Same here, @rdsubhas solution sounds very nice actually. +1 for containerManagementPolicy

**Comment by nmittler** [comment]
@rdsubhas your idea for `containerManagementPolicy` sounds good in general. To clarify/expand a bit, if the policy is ordered:

- Startup: a container has to be `Ready` before the next one starts. This will ensure that Envoy can process outbound requests as soon as the application starts.
- Shutdown: all shutdown hooks should be run in sequence, before terminating containers. This will allow envoy to drain inbound listeners before the application is terminated.

**Comment by xmlking** [comment]
@rdsubhas we have some init-containers that need network access. This makes even worse to use istio as  1st istio-init` is not even updated iptables

**Comment by bjhaid** [comment]
This PR:

https://github.com/kubernetes/enhancements/pull/919

was merged a while ago, should this issue be closed? cc @Joseph-Irving

**Comment by hzxuzhonghu** [comment]
Nope, this should be closed when the feature is implemented.

**Comment by Thoro** [comment]
The PR also suggestions something different, it solves the issue of istio, but it does not solve the ordered startup / shutdown of containers

**Comment by cwen0** [comment]
> I have a different use case where a side car container creates a fuse filesystem that is consumed by the application container. The application container should not run its code until the side car is ready. The side car needs to run for as long as the application pod and should be stopped after the application pod. We would like the application container to not host custom logic for detecting the readiness of the mount. We don not route traffic to the pod but process data inside it. The pod level readiness of each container does not help the cross container dependency.

@kerneltime Hi, I have the same issue when using fuse in Kubernetes. How do you temporarily solve this issue? Thanks!

**Comment by mohithg** [comment]
I have a similar use case, I have a sidecar container which needs to be started before the main container is started because it watches for the changes in the file system, is there a temporary solution to this?

**Comment by galindro** [comment]
And if a would like to run an one-off container that should be executed *after* other container in the same pod? Unfortunately [Container PostStart Lifecycle Hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/?spm=a2c41.12820884.0.0.22345fa8qLBgIw#container-hooks) isn't good because there is no guaranties that it would always been triggered after container's ENTRYPOINT. If I could have a container initialization order + the possibility to start a on-off container, would be nice.

My use case is: I'm deploying a Teleport Auth server and, after its initialization is done, I need to join it in a trusted cluster by executing an one-off command.

**Comment by dapeng09** [comment]
hi folks, any plan on implementing this nice feature or is there a temporary solution for this(long-running containers startup order within a pod)?

**Comment by astleychen** [comment]
> hi folks, any plan on implementing this nice feature or is there a temporary solution for this(long-running containers startup order within a pod)?

if `global.outboundTrafficPolicy.mode` is `ALLOW_ANY ` which is the default, you may try to add below annotation to pod or deployment pod spec so that there is no transient traffic blocking for your containers.

`traffic.sidecar.istio.io/excludeOutboundIPRanges: "0.0.0.0/0"`

**Comment by montanaflynn** [comment]
I want this for adding a [fluentd](https://docs.fluentd.org/deployment/high-availability#network-topology) sidecar container alongside my web servers. The web servers will connect to fluentd upon startup and if it cannot connect exit. Right now I just get a bunch of restarts on the server, which isn't terrible but it does mess with our logging and shows restarts that shouldn't happen in my opinion. Using initContainers isn't an option since the containers are all in the same pod.

The ideal logic would be:

1. Start fluentd with a readiness probe
2. Start server that needs to connect to fluentd after fluentd container is ready

I don't know what the configuration would look like but I think it's a very commonly requested feature in both kubernetes and docker compose so there is for sure interest.

**Comment by Joseph-Irving** [comment]
You can track progress on this feature here https://github.com/kubernetes/enhancements/issues/753

**Comment by krackjack29** [comment]
Is there any plan to add this feature??

**Comment by tariq1890** [comment]
Commenting for thread activity

**Comment by rata** [comment]
@tariq1890 @krackjack29 Yes, as @Joseph-Irving [mentioned][link] this issue is tracking progress of a KEP to address it: https://github.com/kubernetes/enhancements/issues/753. I've just opened a PR to modify the KEP: https://github.com/kubernetes/enhancements/pull/1913

Can I ask (please answer here, not in the other issues/PRs) if your use case to start some containers before others, do you also need those containers to run before `initContainers`, after or is not relevant in your use cases?


[link]: https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-577708535

**Comment by ldemailly** [comment]
after init container as init containers already provide ordering guarantees

**Comment by luksa** [comment]
No, a sidecar KEP shouldn't be considered complete until it allows running sidecars during the init phase, too. For example, pods in a service mesh typically also need the same service mesh functionality for their init containers. And while there are workarounds for proper ordering of the start/shutdown of the pod's main containers, Kubernetes currently has no way of running sidecars in the init phase.

**Comment by sathieu** [comment]
A slightly different approach is needed with Istio with [CNI](https://preliminary.istio.io/latest/docs/setup/additional-setup/cni/). The problem is described [here](https://preliminary.istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers) *Init containers execute before the sidecar proxy starts, which can result in traffic loss during their execution*.

What is needed for that situation is starting envoy before the first initContainer and keep running once Pod is initialized. Maybe a new `earlyContainers` spec?

**Comment by thepip3r** [comment]
As another case we're using and running into this issue:

We have an application that the vendor claims requires it's own SQL instance (not database, but instance of SQL).  So we're using a multi-container deployment as mssql+application.  Because there are n-instances of that multi-container pod, the application logic checks mssql to see if the database is already present; if so, it binds to it; if not, it assumes it's a new instance and creates a new database.  Initially this worked fine.  However, over time, due to race-conditions, we ran into situations where MSSQL wasn't presenting 1433 before the application attempted the bind and in the delta between the initial check and the prepping for the database schema creation, sql would come up so the application would effectively wipe the databases during the rebuild.  We got around this issue by adding code to the application container that looped to check for 1433 before attempting to check to see if the database exists but it would be very valuable to us if we could make the application container depend on the mssql container's successful startup.

**Comment by rata** [comment]
@thepip3r cool, thanks! Please see the google doc linked in [this email thread][link] and add the use case there!

[link]: https://groups.google.com/g/kubernetes-dev/c/pY6AOwIgs8Y/m/Ptymao18CwAJ

**Comment by CSimpiFoN** [comment]
Hello, another use case, that could benefit from if this gets implemented. Our microservices get their configuration and BE credentials from Hashicorp Vault through a sidecar, that manages the communication with Vault. It would be great if we can ensure that this sidecar starts up before the main application does, so it can get the necessary secrets right after startup.

**Comment by rata** [comment]
@CSimpiFoN Please see [this message][link] with links on how to add the use case to the use cases doc.


[link]: https://github.com/kubernetes/kubernetes/issues/65502#issuecomment-889158054

**Comment by nikskiz** [comment]
> Hello, another use case, that could benefit from if this gets implemented. Our microservices get their configuration and BE credentials from Hashicorp Vault through a sidecar, that manages the communication with Vault. It would be great if we can ensure that this sidecar starts up before the main application does, so it can get the necessary secrets right after startup.

@CSimpiFoN Depending on the reason why you need a sidecar, there could be an alternative to your use case. You can use mutating webhooks that will inject secrets into your pod before your main application starts up. 

https://banzaicloud.com/docs/bank-vaults/mutating-webhook/

**Comment by newwuhan5** [comment]
We have exactly same issue when using sidecars with Envoy. The sidecars are not ready, but Envoy already marks as healthy and starts to take the traffics, which results in the 500 errors.  We have a filter of Envoy calling gRPC service from sidecar for jwt resigning.  Is there a decent solution or scripting example from someone?  Or any progress on the different solutions described above?

**Comment by newwuhan5** [comment]
@rata, Thank you.  It looks like an elegant solution, which we will explore if we could add to our Envoy pod.

**Comment by newwuhan5** [comment]
@Jeskz0rd Thank you for confirming that you team has been using kubexit.

**Comment by neoakris** [comment]
In many cases this trick can be used to workaround the issue.
https://stackoverflow.com/questions/33887194/how-to-set-multiple-commands-in-one-yaml-file-with-kubernetes

**Comment by ringerc** [comment]
`kubeexit` is the inner system effect in full force, eh? Necessary to work around the current limitations, but pretty unfortunate that it's required.

Lots of these issues can be worked around with a mix of application coding ("if expected sidecar not responsive, retry loop") and the use of readiness checks to prevent requests from being routed to pods before all their containers are up. But it's ugly for sure.

**Comment by lavender2020** [comment]
We also have have the same issues with istio-proxy and application containers in one pod.
Have to start istio-proxy firstly. And after that, you can then start up application containers.

**Comment by chadlwilson** [comment]
@lavender2020 Istio since `1.7` already has its own magic to workaround this (for start-up, not for shutdown order to my knowledge) via `holdApplicationUntilProxyStarts` in mesh `ProxyConfig`: https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/

**Comment by dims** [comment]
Let's use the following enhancement issue: https://github.com/kubernetes/enhancements/issues/753

Here's the plan in 1.28: https://github.com/kubernetes/enhancements/pull/3968

/close
